File: src/java/spring/src/main/java/app/planentnine/springsessionless/application/util/Sessionless.java
package app.planentnine.springsessionless.application.util;

import org.apache.commons.lang3.StringUtils;
import org.bouncycastle.crypto.params.ECDomainParameters;
import org.bouncycastle.crypto.params.ECPrivateKeyParameters;
import org.bouncycastle.crypto.params.ECPublicKeyParameters;
import org.bouncycastle.crypto.signers.ECDSASigner;
import org.bouncycastle.jcajce.provider.digest.Keccak;
import org.bouncycastle.jce.ECNamedCurveTable;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.jce.spec.ECNamedCurveParameterSpec;

import java.math.BigInteger;
import java.security.InvalidAlgorithmParameterException;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.NoSuchProviderException;
import java.security.SecureRandom;
import java.security.Security;
import java.security.interfaces.ECPrivateKey;
import java.security.interfaces.ECPublicKey;
import java.security.spec.ECPoint;
import java.util.UUID;

//TODO: upload artifact and reimport
public class Sessionless {
    
    public static String[] generateKeysAsHex() {
        KeyPairGenerator generator;
        Security.addProvider(new BouncyCastleProvider());
        ECNamedCurveParameterSpec ecNamedCurveParameterSpec =
                ECNamedCurveTable.getParameterSpec("secp256k1");
        
        try {
            generator = KeyPairGenerator.getInstance("ECDSA", "BC");
            generator.initialize(ecNamedCurveParameterSpec, new SecureRandom());
            
        } catch (NoSuchAlgorithmException | NoSuchProviderException | InvalidAlgorithmParameterException e) {
            throw new RuntimeException(e);
        }
        
        
        KeyPair keyPair = generator.generateKeyPair();
        ECPrivateKey ecPrivateKey = (ECPrivateKey) keyPair.getPrivate();
        ECPublicKey ecPublicKey = (ECPublicKey) keyPair.getPublic();
        
        String privateKeyHex = extractPrivateKeyHex(ecPrivateKey);
        String publicKeyHex = extractPublicKeyHex(ecPublicKey);
        
        return new String[]{privateKeyHex, publicKeyHex};
    }
    
    public static String[] signMessage(String privateKey, String message) {
        ECNamedCurveParameterSpec ecNamedCurveParameterSpec =
                ECNamedCurveTable.getParameterSpec("secp256k1");
        ECDomainParameters domain =
                new ECDomainParameters(
                        ecNamedCurveParameterSpec.getCurve(),
                        ecNamedCurveParameterSpec.getG(),
                        ecNamedCurveParameterSpec.getN(),
                        ecNamedCurveParameterSpec.getH());
        
        BigInteger privateKeyFormatted = new BigInteger(privateKey, 16);
        
        ECPrivateKeyParameters privateKeyParameters = new ECPrivateKeyParameters(privateKeyFormatted, domain);
        ECDSASigner signer = new ECDSASigner();
        signer.init(true, privateKeyParameters);
        
        byte[] messageHash = keccakMessageHash(message);
        BigInteger[] signature = signer.generateSignature(messageHash);
        
        return new String[]{signature[0].toString(16), signature[1].toString(16)};
    }
    
    public static boolean verify(String publicKey, String[] signature, String message) {
        BigInteger publicKeyFormatted = new BigInteger(publicKey, 16);
        ECNamedCurveParameterSpec ecNamedCurveParameterSpec =
                ECNamedCurveTable.getParameterSpec("secp256k1");
        ECDomainParameters domain =
                new ECDomainParameters(
                        ecNamedCurveParameterSpec.getCurve(),
                        ecNamedCurveParameterSpec.getG(),
                        ecNamedCurveParameterSpec.getN(),
                        ecNamedCurveParameterSpec.getH());
        org.bouncycastle.math.ec.ECPoint publicKeyPoint = ecNamedCurveParameterSpec.getCurve().decodePoint(publicKeyFormatted.toByteArray());
        ECPublicKeyParameters publicKeyParameters = new ECPublicKeyParameters(publicKeyPoint, domain);
        
        ECDSASigner signer = new ECDSASigner();
        signer.init(false, publicKeyParameters);
        
        MessageDigest digest = new Keccak.Digest256();
        byte[] messageHash = digest.digest(message.getBytes());
        
        return signer.verifySignature(messageHash, new BigInteger(signature[0], 16), new BigInteger(signature[1], 16));
    }
    
    public static boolean associate(String primaryPublicKey, String[] primarySignature, String primaryMessage,
                                    String secondaryPublicKey, String[] secondarySignature, String secondaryMessage) {
        return verify(primaryPublicKey, primarySignature, primaryMessage)
                && verify(secondaryPublicKey, secondarySignature, secondaryMessage);
    }
    
    public static UUID generateUuid() {
        return UUID.randomUUID();
    }
    
    private static String extractPrivateKeyHex(ECPrivateKey ecPrivateKey) {
        BigInteger privateKeyBigInt = ecPrivateKey.getS();
        return privateKeyBigInt.toString(16);
    }
    
    private static String extractPublicKeyHex(ECPublicKey ecPublicKey) {
        
        ECPoint ecPoint = ecPublicKey.getW();
        BigInteger publicBytesRawX = ecPoint.getAffineX();
        BigInteger pubicBytesRawY = ecPoint.getAffineY();
        
        //Add compression prefix based on sign
        boolean yIsEven = pubicBytesRawY.mod(new BigInteger("2")).equals(BigInteger.ZERO);
        publicBytesRawX = publicBytesRawX.abs();
        String prefix = yIsEven ? "02" : "03";
        
        //Ensure stripped 0's are sign only
        String publicKeyHex = publicBytesRawX.toString(16);
        publicKeyHex = StringUtils.leftPad(publicKeyHex, 64, '0');
        publicKeyHex = prefix + publicKeyHex;
        
        return publicKeyHex;
    }
    
    private static byte[] keccakMessageHash(String message) {
        MessageDigest digest = new Keccak.Digest256();
        return digest.digest(message.getBytes());
    }
}

File: src/kotlin/src/main/kotlin/com/planetnine/sessionless/util/sessionless/impl/Sessionless.kt
package com.planetnine.sessionless.util.sessionless.impl

import com.planetnine.sessionless.util.sessionless.impl.Sessionless.WithCustomVault
import com.planetnine.sessionless.util.sessionless.impl.Sessionless.WithKeyStore
import com.planetnine.sessionless.util.sessionless.models.ISessionless
import com.planetnine.sessionless.util.sessionless.models.KeyAccessInfo
import com.planetnine.sessionless.util.sessionless.models.MessageSignature
import com.planetnine.sessionless.util.sessionless.models.SimpleKeyPair
import com.planetnine.sessionless.util.sessionless.models.vaults.ICustomVault
import com.planetnine.sessionless.util.sessionless.models.vaults.IVault
import com.planetnine.sessionless.util.sessionless.util.KeyUtils
import com.planetnine.sessionless.util.sessionless.util.KeyUtils.domainParameters
import com.planetnine.sessionless.util.sessionless.util.KeyUtils.toECHex
import com.planetnine.sessionless.util.sessionless.util.KeyUtils.toECPrivateKey
import com.planetnine.sessionless.util.sessionless.util.KeyUtils.toHex
import com.planetnine.sessionless.util.sessionless.util.hashKeccak256
import org.bouncycastle.crypto.params.ECPrivateKeyParameters
import org.bouncycastle.crypto.params.ECPublicKeyParameters
import org.bouncycastle.crypto.signers.ECDSASigner
import java.math.BigInteger
import java.security.KeyPair
import java.security.PrivateKey
import java.security.interfaces.ECPrivateKey
import java.util.Base64
import java.util.UUID

/** [Sessionless] implementation (sealed, check the subclasses)
 * @see WithKeyStore
 * @see WithCustomVault */
sealed class Sessionless(override val vault: IVault) : ISessionless {
    /** [ISessionless.WithKeyStore] implementation with [Sessionless] as superclass */
    class WithKeyStore(override val vault: KeyStoreVault) :
        Sessionless(vault),
        ISessionless.WithKeyStore {

        override fun generateKeys(keyAccessInfo: KeyAccessInfo): KeyPair {
            val pair = KeyUtils.generateKeyPair()
            vault.save(pair, keyAccessInfo)
            return pair
        }

        override suspend fun generateKeysAsync(keyAccessInfo: KeyAccessInfo): KeyPair {
            val pair = KeyUtils.generateKeyPairAsync()
            vault.save(pair, keyAccessInfo)
            return pair
        }

        override fun getKeys(keyAccessInfo: KeyAccessInfo): KeyPair {
            return vault.get(keyAccessInfo)
        }

        override fun sign(message: String, keyAccessInfo: KeyAccessInfo): MessageSignature {
            val privateKey = getKeys(keyAccessInfo).private
            return sign(message, privateKey)
        }
    }

    /** [ISessionless.WithCustomVault] implementation with [Sessionless] as superclass */
    class WithCustomVault(override val vault: ICustomVault) :
        Sessionless(vault),
        ISessionless.WithCustomVault {

        override fun generateKeys(): SimpleKeyPair {
            val pair = KeyUtils.generateKeyPair()
            val simple = pair.toECHex()
            vault.save(simple)
            return simple
        }

        override suspend fun generateKeysAsync(): SimpleKeyPair {
            val pair = KeyUtils.generateKeyPairAsync()
            val simple = pair.toECHex()
            vault.save(simple)
            return simple
        }

        override fun getKeys(): SimpleKeyPair {
            return vault.get()
        }

        override fun sign(message: String): MessageSignature {
            val privateString = getKeys().privateKey
            val privateKey = privateString.toECPrivateKey(KeyUtils.Defaults.parameterSpec)
            return sign(message, privateKey)
        }
    }

    override fun sign(message: String, privateKey: PrivateKey): MessageSignature {
        val signer = ECDSASigner().apply {
            val privateHex = (privateKey as ECPrivateKey).toHex()
            val privateKeyFormatted = BigInteger(privateHex)
            val privateKeyParameters = ECPrivateKeyParameters(
                privateKeyFormatted,
                KeyUtils.Defaults.parameterSpec.domainParameters
            )
            init(true, privateKeyParameters)
        }
        val messageHash = hashKeccak256(message)
        val signature = signer.generateSignature(messageHash)
        return MessageSignature.from(signature)!!
    }

    override fun verify(
        publicKey: String,
        signature: MessageSignature,
        message: String
    ): Boolean {
        val signer = ECDSASigner().apply {
            val publicBytes = Base64.getDecoder().decode(publicKey)
            val publicInt = BigInteger(publicBytes)
            val paramSpec = KeyUtils.Defaults.parameterSpec
            val publicKeyPoint = paramSpec.curve.decodePoint(publicInt.toByteArray())
            val publicKeyParameters = ECPublicKeyParameters(
                publicKeyPoint, paramSpec.domainParameters
            )
            init(false, publicKeyParameters)
        }
        val messageHash = hashKeccak256(message)
        return signer.verifySignature(
            messageHash, signature.r, signature.s
        )
    }

    override fun generateUUID(): String {
        return UUID.randomUUID().toString()
    }

    override fun associate(
        primaryPublicKey: String,
        primaryMessage: String,
        primarySignature: MessageSignature,
        secondaryPublicKey: String,
        secondaryMessage: String,
        secondarySignature: MessageSignature,
    ): Boolean {
        val verified1 = verify(primaryPublicKey, primarySignature, primaryMessage)
        val verified2 = verify(secondaryPublicKey, secondarySignature, secondaryMessage)
        return verified1 && verified2
    }
}